### PR TITLE
Add DSA indexer weights projection precision controls

### PR DESCRIPTION
**Pull request by buptzyb**
## Related PRs

- Dev counterpart: #6279
- TileLang DSA base: #5049 (already merged into `main`)

## Summary

- Add independent controls for whether the ordinary DSA indexer `linear_weights_proj` participates in FP8/FP4 quantization and whether its output is BF16 or FP32.
- Preserve the existing defaults: follow the enclosing quantization context and return BF16 projection output.
- Add a non-quantized true-FP32-output path using BF16 parameters and inputs, Transformer Engine `general_gemm` when supported, and a cached FP32 `torch.mm` fallback otherwise.
- Preserve checkpoint keys and BF16 parameter storage while retaining immediate and deferred WGrad, gradient-accumulation fusion, FSDP main-grad buffers, CUDA Graph capture, and asynchronous parameter readiness.
- Keep the cuDNN DSA backend on its required BF16 weights-tensor contract; FP32 projection output remains available to the unfused and TileLang paths.

## Testing

- Megatron-LM autoformat check against `main`: passed with Black 26.3, isort, Pylint 10/10, and Ruff.
- Python 3 static compilation of the changed source and precision test files: passed.
- Corresponding #6279 precision suite: 33 passed / 4 skipped on H100 and 35 passed / 1 skipped on GB200; the delayed + fused-WGrad hook regression passed.
- Added main-specific regressions for asynchronous projection-weight readiness and the `DSAIndexer` deferred-WGrad flush chain.

### CODE REVIEW

**Review by zhongbozhu — APPROVED** [review]

File: megatron/core/distributed/fsdp/mcore_fsdp_adapter.py [hunk 1]
```diff
@@ -99,6 +99,16 @@ def _fine_grained_recurse_module_types(
         config: TransformerConfig, ddp_config: DistributedDataParallelConfig
     ) -> Tuple[Type[nn.Module], ...]:
         """Module classes needing ``parameters(recurse=True)`` for fine-grained hooks."""
+        recurse_types: List[Type[nn.Module]] = []
+
+        if config.dsa_indexer_weights_proj_output_dtype == "fp32":
+            # DSAttention calls DSAIndexer.forward_before_topk directly, bypassing the indexer's
+            # module hooks and the weights projection's module call. Gather its nested parameters
+            # at the DSAttention boundary for both forward and backward.
+            from megatron.core.transformer.experimental_attention_variant.dsa import DSAttention
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/mcore_fsdp_adapter.py]: move to top-level

File: megatron/core/distributed/fsdp/mcore_fsdp_adapter.py [hunk 2]
```diff
@@ -99,6 +99,16 @@ def _fine_grained_recurse_module_types(
         config: TransformerConfig, ddp_config: DistributedDataParallelConfig
     ) -> Tuple[Type[nn.Module], ...]:
         """Module classes needing ``parameters(recurse=True)`` for fine-grained hooks."""
+        recurse_types: List[Type[nn.Module]] = []
+
+        if config.dsa_indexer_weights_proj_output_dtype == "fp32":
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/mcore_fsdp_adapter.py:170]: Why?

File: megatron/core/distributed/fsdp/mcore_fsdp_adapter.py (hunk 1 above)
Comment (buptzyb) [reply megatron/core/distributed/fsdp/mcore_fsdp_adapter.py]: done

File: megatron/core/distributed/fsdp/mcore_fsdp_adapter.py (hunk 2 above)
Comment (buptzyb) [reply megatron/core/distributed/fsdp/mcore_fsdp_adapter.py:170]: Only the FP32 path needs the recursive boundary because it passes the projection weight directly to custom autograd, bypassing the linear module's FSDP hooks. `ensure_params_ready` covers the forward read but does not provide a pre-backward unshard hook. The BF16 path calls the module normally, so its leaf hooks handle both passes. I updated the comment to clarify this distinction.
